src/ManageCourses.Api/Services/Data/DataService.cs

```diff
@@ -42,109 +42,64 @@ public void ProcessUcasPayload(UcasPayload payload)
 
             _logger.LogWarning("Beginning UCAS import");
             _logger.LogInformation($"Upserting {allInstitutions.Count()} institutions");
-            int processed = 0;
-            foreach (var inst in allInstitutions)
+            using (var transaction = (_context as DbContext).Database.BeginTransaction())
             {
-                // note: this does not perform a delete of existing institutions
-
-                var instRes = AttemptTransaction(inst.InstCode, "upsert institution entity", () => UpsertInstitution(inst));
-
-                var deleted = AttemptTransaction(inst.InstCode, "delete for replace", () => DeleteForInstitution(inst.InstCode));
-                var added = AttemptTransaction(inst.InstCode, "insert", () => AddForInstitution(inst.InstCode, payload));
-
-                if (added == null && deleted != null)
+                try 
                 {
-                    //deletion succeeded but adding didn't... attempt to recover by re-adding deleted content
-                    AttemptTransaction(inst.InstCode, "restore", () => AddForInstitution(inst.InstCode, deleted));
+                    foreach(var inst in allInstitutions)
```

Comment:
Change `UpsertInstitution` to `UpsertInstitutions`
Return a list of Institution that was either newly added or updated as `iAmNowInContext`
Looks like its missing 
`_context.Save()`

Anyways then in `AddUcasData(payload, iAmNowInContext)`

And do the match up of institution via `ref `